cpp/src/iterator/iterator.cuh

```diff
@@ -0,0 +1,459 @@
+ * @brief provide column input iterator with/without nulls
+ * @file iterator.cuh
+ *
+ * This column input iterator is designed to be able to be used as input
+ * iterator for thrust and cub.
+ *
+ * The input iterator is implemented using thrust::iterator_adaptor
+ * and thrust::counting_iterator. When creating a null supported input iterator,
+ * the iterator requires pointer of data and null bitmap and an identity value
+ * like below.
+ *
+ * bool has_nulls = true;
+ * auto it_dev = cudf::make_iterator<has_nulls, T>(column, T{0});
+ *
+ * (it_dev +id) returns static_cast<T*>( column->data )[id] if column->valid
+ * at id is true, and T{0} if column->valid at id = false.
+ *
+ * The identity value depends on the data type and the aggregation type.
+ * e.g.
+ * T = int32_t and aggregation is a kind of `sum`:
+ *     identity = int32_t{0}
+ * T = cudf::date32  and aggregation is a kind of `max`:
+ *     identity = std::numeric_limits<cudf::date32>::lowest()
+ *
+ * Examples of template parameter:
+ * 1. template parameter for upcasting input iterator
+ *     make_iterator<has_nulls, T, T_upcast>(...)
+ *
+ * 2. template parameter for upcasting + squared input iterator
+ *     using T_mutator = cudf::detail::mutator_squared<T_upcast>;
+ *     make_iterator<has_nulls, T, T_upcast, T_mutator>(...)
+ *
+ * 3. template parameter for using `mutator_meanvar`
+ *     using T_output = cudf::detail::mutator_meanvar<T_upcast>
+ *     make_iterator<has_nulls, T, T_output, T_output>(...)
+ *
+ * 4. template parameter for custom indexed iterator
+ *     using out_helper = cudf::detail::mutator_single<T>;
+ *     using T_index = gdf_index_type;
+ *     make_iterator<has_nulls, T, T, out_helper, T_index*>(...)
+ * -------------------------------------------------------------------------**/
+
+#ifndef CUDF_ITERATOR_CUH
+#define CUDF_ITERATOR_CUH
+
+#include <cudf.h>
+#include <bitmask/bit_mask.cuh>         // need for bit_mask::bit_mask_t
+#include <utilities/cudf_utils.h>       // need for CUDA_HOST_DEVICE_CALLABLE
+#include <utilities/error_utils.hpp>
+#include <utilities/type_dispatcher.hpp>
+
+#include <thrust/iterator/iterator_adaptor.h>
+#include <thrust/iterator/counting_iterator.h>
+
+namespace cudf
+{
+namespace detail
+{
+// ---------------------------------------------------------------------------
+// helper structs of output for column_input_iterator
+
+/** -------------------------------------------------------------------------*
+ * @brief mutator for output a scalar value
+ * A mutator for `column_input_iterator`
+ * `column_input_iterator` creates this struct with parameter `_value`, `is_valid`
+ *
+ * This struct computes the output value as `static_cast<T>(_value)`.
+ *
+ * @tparam  T a scalar data type to be output
+ * @tparam  T_input a scalar data type to be input
+ * -------------------------------------------------------------------------**/
+template<typename T>
+struct mutator_single
```

Comment:
What still bothers me about the `mutator` objects is that there is a expected interface that all `mutator` objects should implement, but currently there's no description nor enforcement of that interface.  

For example, every `mutator` must have a constructor template with a signature like `(T, bool)`, or an implicit conversion operator to `T`. 

Obviously we don't want to use an abstract class with pure-virtual functions to define an interface. We could maybe use CRTP to define and enforce an interface. But at the very least, the `mutator` objects warrant a thorough comment block describing what they are for and their expected interface.